simulator/ssh/client.go

```diff
@@ -0,0 +1,211 @@
+// Package ssh provides just enough SSH/SFTP functionality to perform random writes
+// to an SFTP server.
+package ssh
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"time"
+
+	sdipacket "github.com/alphasoc/sftp-data-input/packet"
+	sdifxp "github.com/alphasoc/sftp-data-input/packet/fxp"
+
+	"golang.org/x/crypto/ssh"
+)
+
+// SSH/SFTP client version.  We don't perform any version negotiation.
+const ClientVer = 3
+
+// Only SSH_FX_OK error code is needed for our purposes.
+const SSH_FX_OK = 0
+
+// SSH/SFTP requests require an ID.  Per
+// https://datatracker.ietf.org/doc/html/draft-ietf-secsh-filexfer-13#section-4, they
+// don't need to be unique.  Since we're not sending requests/data in parallel, an id of
+// 1 is just fine.
+const reqID = 1
+
+// Client wraps SSH client and session structs.
+type Client struct {
+	sshClient *ssh.Client
+	sess      *ssh.Session
+	w         io.WriteCloser
+	r         io.Reader
+}
+
+// NewClient initializes and returns a Client ready to be used for SSH/SFTP transfer along
+// with an error.  Note that Teardown should be called when the Client is no longer needed.
+func NewClient(user string, src net.IP, dst string, signer ssh.Signer) (*Client, error) {
+	// ClientConfig will use pubkey auth, ignore the host key and apply a 5 second connection
+	// timeout.
+	config := &ssh.ClientConfig{
+		User: user,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(signer),
+		},
+		// TODO insecure ok?
+		//HostKeyCallback: ssh.FixedHostKey(hostKey),
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+		Timeout:         5 * time.Second,
+	}
+	// Create dialer with src address, dial and setup a new client.
+	d := &net.Dialer{
+		LocalAddr: &net.TCPAddr{IP: src},
+		Timeout:   config.Timeout,
+	}
+	nConn, err := d.Dial("tcp", dst)
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect: %v", err)
+	}
+	sshConn, chans, reqs, err := ssh.NewClientConn(nConn, dst, config)
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect: %v", err)
```

Comment:
Will nnConn be closed in case of an error?

And the end I think the only resource acquired here is a connection, so maybe we don't need to close the session and client and can focus on nConn only? (this would need to be verified). Also, probably this could be refactored a little bit so we can always force closing via defer in case of an error, but might be not worth the effort.

Reply:
I'll verify and get back to you.

Reply:
OK, some findings.

Closing the ssh.Client does indeed close the underlying nConn.
Closing the ssh.Session has the added benefit of also closing the session channel (used by the read/write pipes).

Seems to me like this is a good approach.  Do let me know, however, if you think we should refactor.

Reply:
Sure, the question is whether nConn is closed in case NewClientConn returns error (as then you don't get sshConn and don't close the connection).

Reply:
It does.  2 potential error cases in ssh.NewClientConn... both branches close out the passed in nConn.